tests/main/download-timeout/task.yaml

```diff
@@ -0,0 +1,38 @@
+summary: Test that downloads time-out on very slow connection.
+
+details: |
+  Test that snap download times out on a very slow connection. Download bandwidth
+  is reduced with 'tc', and in addition the download parameters for snapd are
+  tweaked for the test.
+
+systems: [ubuntu-18.04*, ubuntu-2*]
+
+prepare: |
+  cp /etc/systemd/system/snapd.service.d/local.conf snapd.local.conf
+  # set minimum download speed that is a bit higher than the rate-limited speed set with tc.
+  # also force 1-minute window, so the test doesn't have to run for full 5 minutes (the default
+  # window).
+  sed -i 's/\(Environment=.*\)/\1 SNAPD_MIN_DOWNLOAD_SPEED=99000 SNAPD_DOWNLOAD_MEAS_WINDOW=1m/' /etc/systemd/system/snapd.service.d/local.conf
+  systemctl daemon-reload
+  systemctl restart snapd.{socket,service}
+
+execute: |
+  tc qdisc add dev ens4 ingress
+  tc filter add dev ens4 root protocol ip u32 match u32 0 0 police rate 32kbit burst 16k drop flowid :1
+  tc filter add dev ens4 parent ffff: protocol ip u32 match u32 0 0 police rate 32kbit burst 16k drop flowid :1
+
+  echo "Installing a large snap fails if connection is very slow"
+  snap install lxd 2>&1 | MATCH "download too slow:"
+
+  echo "Downloading a large snap fails too"
+  SNAP_MIN_DOWNLOAD_SPEED=99000 SNAP_DOWNLOAD_MEAS_WINDOW=1m snap download lxd 2>&1 | MATCH "download too slow:"
```

Comment:
SNAP_MIN_DOWNLOAD_SPEED and SNAP_DOWNLOAD_MEAS_WINDOW could be set in the test `environment:` section

Reply:
Yes, I'm just not sure if it is not going to be slightly confusing if they affect only 'snap download', but snapd still needs the sed hack?

Reply:
Yes, you'd still need the hack, but it can be made simpler:
```
sed -i "s/\(Environment=.*\)/\1 SNAPD_MIN_DOWNLOAD_SPEED=${SNAPD_MIN_DOWNLOAD} SNAPD_DOWNLOAD_MEAS_WINDOW=${SNAPD_DOWNLOAD_MEAS_WINDOW}/"
```

In the long run, maybe it'd be better if the tests added something like `/etc/systemd/system/snapd.service.d/env-overrides.conf` which only contains Environment overrides and we could just append to it, eg.:
```
cat <<EOF >> /etc/systemd/system/snapd.service.d/env-overrides.conf
Environment=SNAP_MIN_DOWNLOAD_SPEED=${SNAP_MIN_DOWNLOAD_SPEED}
Environment=SNAP_DOWNLOAD_MEAS_WINDOW=${SNAP_DOWNLOAD_MEAS_WINDOW}
EOF
```

Reply:
That's a great idea, thanks! Updated.